src/dlg_audiooptions.cpp

```diff
@@ -988,143 +798,112 @@ void AudioOptsDialog::UpdatePlot(PlotScalar *plotScalar)
+    SRC_STATE          *src;
+    FIFO               *fifo, *callbackFifo;
+    int src_error;
+    
+    fifo = codec2_fifo_create((int)(DT*TEST_WAVEFORM_PLOT_FS*2)); assert(fifo != NULL);
+    src = src_new(SRC_SINC_FASTEST, 1, &src_error); assert(src != NULL);
+    
+    auto engine = AudioEngineFactory::GetAudioEngine();
+    auto devList = engine->getAudioDeviceList(IAudioEngine::IN);
+    for (auto& devInfo : devList)
+    {
+        if (wxString(devInfo.name).Trim() == devName.Trim())
         {
-            Pa_ReadStream(stream, in48k_stereo_short, TEST_BUF_SIZE);
-            if (inputChannels == 2) {
-                for(j=0; j<TEST_BUF_SIZE; j++)
-                    in48k_short[j] = in48k_stereo_short[2*j]; // left channel only
-            }
-            else {
-                for(j=0; j<TEST_BUF_SIZE; j++)
-                    in48k_short[j] = in48k_stereo_short[j]; 
-            }
-            int n8k = resample(src, in8k_short, in48k_short, 8000, sampleRate, TEST_BUF_SIZE, TEST_BUF_SIZE);
-            resample_for_plot(fifo, in8k_short, n8k, FS);
-
-            short plotSamples[TEST_WAVEFORM_PLOT_BUF];
-            if (codec2_fifo_read(fifo, plotSamples, TEST_WAVEFORM_PLOT_BUF))
+            int sampleCount = 0;
+            int sampleRate = wxAtoi(m_cbSampleRateRxIn->GetValue());
+            auto device = engine->getAudioDevice(
+                devInfo.name, 
+                IAudioEngine::IN, 
+                sampleRate,
+                devInfo.maxChannels >= 2 ? 2 : 1);
+            
+            if (device)
             {
-                // come back when the fifo is refilled
-                continue;
-            }
-
-            plotScalar->add_new_short_samples(0, plotSamples, TEST_WAVEFORM_PLOT_BUF, 32767);
-            sampleCount += TEST_WAVEFORM_PLOT_BUF;
-            CallAfter(&AudioOptsDialog::UpdatePlot, plotScalar);
-        }
+                std::mutex callbackFifoMutex;
+                std::condition_variable callbackFifoCV;
+            
+                callbackFifo = codec2_fifo_create(sampleRate);
+                assert(callbackFifo != nullptr);
+                
+                device->setOnAudioData([&](IAudioDevice&, void* data, size_t numSamples, void* state) {
+                    short* in48k_stereo_short = static_cast<short*>(data);
+                    short               in48k_short[numSamples];
+                
+                    if (devInfo.maxChannels >= 2) {
+                        for(size_t j = 0; j < numSamples; j++)
+                            in48k_short[j] = in48k_stereo_short[2*j]; // left channel only
+                    }
+                    else {
+                        for(size_t j = 0; j < numSamples; j++)
+                            in48k_short[j] = in48k_stereo_short[j]; 
+                    }
+                
+                    {
+                        std::unique_lock<std::mutex> callbackFifoLock(callbackFifoMutex);
```

Comment:
might not need this mutex?

Reply:
Actually, as the Play/Rec 2s buttons are implemented now, it is necessary in order for the plots to render correctly. It might be worth rethinking their implementation, though.